### ISSUE TITLE
test: add line-search contract tests on non-trivial problems

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#62]
## Problem

The existing line-search tests in `test_line_search.py` only verify the Armijo/Wolfe conditions on a trivial 1D quadratic `phi(alpha) = 0.5*(1-alpha)^2`. There are no tests verifying that:

1. Armijo backtracking returns `alpha < alpha0` (it actually shrinks the step).
2. Strong Wolfe returns `alpha > 0` with `dphi_alpha` having the correct sign.
3. The line search works correctly when integrated into Newton/LM on a non-trivial (e.g., Rosenbrock-style) problem.
4. The line search handles edge cases: very steep vs very flat regions, non-convex objectives, etc.

The progress tests for Newton+line_search and LM+line_search only check that loss decreases on a simple 2D quadratic — this doesn't stress the line search at all since the full Newton step already converges.

## Proposal

Add targeted contract tests:

- `test_armijo_backtracking_shrinks_step_on_steep_function`
- `test_strong_wolfe_returns_positive_alpha`
- `test_newton_line_search_converges_on_rosenbrock`
- `test_lm_line_search_converges_on_nonlinear_residuals`

These would catch regressions in the line-search implementation itself, independent of the optimizer logic.

**Comment by kovacoj** [comment]
## Proposed fix

Add contract tests for the line-search functions on non-trivial problems:

```python
import torch
from optimizers import line_search


def _rosenbrock_phi_factory():
    """Rosenbrock-like function: phi(alpha) = (1 - (1+alpha))**2 + 100*((1+alpha)**2 - (1+2*alpha))**2"""
    # Simplified: use a cubic for step-shrinking behavior
    # phi(alpha) = alpha**3 - 2*alpha**2 + 2  (minimum at alpha=0, local max around alpha~1.33)
    def phi(alpha):
        a = alpha.item() if isinstance(alpha, torch.Tensor) else alpha
        return torch.tensor(a**3 - 2*a**2 + 2.0)
    def dphi(alpha):
        a = alpha.item() if isinstance(alpha, torch.Tensor) else alpha
        return torch.tensor(3*a**2 - 4*a)
    phi0 = phi(torch.tensor(0.0))
    dphi0 = dphi(torch.tensor(0.0))
    return phi, dphi, phi0, dphi0


def test_armijo_backtracking_returns_step_smaller_than_initial():
    phi, dphi, phi0, dphi0 = _rosenbrock_phi_factory()
    alpha0 = torch.tensor(5.0)
    alpha, phi_alpha = line_search.armijo_backtracking(
        phi=phi, phi0=phi0, dphi0=dphi0, alpha0=alpha0,
    )
    assert alpha < alpha0  # must have shrunk from the too-large initial step


def test_strong_wolfe_returns_positive_alpha():
    phi, dphi, phi0, dphi0 = _rosenbrock_phi_factory()
    alpha, phi_alpha, dphi_alpha = line_search.strong_wolfe(
        phi=phi, dphi=dphi, phi0=phi0, dphi0=dphi0,
        alpha0=torch.tensor(5.0),
    )
    assert alpha > 0
    assert bool(phi_alpha <= phi0 + 1e-4 * alpha * dphi0)
    assert bool(torch.abs(dphi_alpha) <= 0.9 * torch.abs(dphi0))


def test_newton_armijo_converges_on_rosenbrock():
    x = torch.nn.Parameter(torch.tensor([-1.0, 1.0]))
    optimizer = Newton([x], line_search_method="armijo")

    def rosenbrock():
        return (1 - x[0])**2 + 100*(x[1] - x[0]**2)**2

    for _ in range(50):
        optimizer.step(rosenbrock)

    with torch.no_grad():
        assert rosenbrock().item() < 1.0


def test_newton_wolfe_converges_on_rosenbrock():
    x = torch.nn.Parameter(torch.tensor([-1.0, 1.0]))
    optimizer = Newton([x], line_search_method="wolfe")

    def rosenbrock():
        return (1 - x[0])**2 + 100*(x[1] - x[0]**2)**2

    for _ in range(50):
        optimizer.step(rosenbrock)

    with torch.no_grad():
        assert rosenbrock().item() < 1.0
```

**Design choices:**
- The cubic test function exercises the step-shrinking behavior that the current quadratic test never triggers.
- Rosenbrock is the standard non-trivial test for second-order methods — it forces the optimizer to navigate a curved valley.
- The convergence threshold (`< 1.0`) is intentionally loose — we're testing that the line search helps, not that Newton solves Rosenbrock in 50 steps.
